### PR TITLE
Add `chrome://` schema URLs to the badge block list

### PR DESCRIPTION
**Pull request by jon-betts**
This is mostly to ensure we aren't serving a lot of content for new tabs in chrome. This should:

 * Return 0
 * Return caching headers

For the moment we are allowing 'chrome-extension://' but I'm not sure if we want to.

To test:

 * `make dev`
 * Goto: http://localhost:5000/api/badge?uri=chrome://newtab
 *  This should return `{"count": 0}`
 * The header 'Cache-Control' should be set to public and have a max-age

This should mean browsers cache the response for a day and don't ask us for it again

### CODE REVIEW

File: h/views/badge.py [hunk 1]
```diff
@@ -22,25 +22,40 @@ def _has_uri_ever_been_annotated(db, uri):
...
+        rf"(?:http[sx]?:)?//(?:{_regex_or_clause(BLOCKED_DOMAINS)})(?:/|$)"
+    )
+
+    # Just the bad schemes
+    _SCHEME_PATTERN = rf"(?:{_regex_or_clause(BLOCKED_SCHEMES)})://"
+
+    # A combo pattern of the bad schemes and the bad domains
+    _PATTERN = re.compile(
+        rf"^(?:{_SCHEME_PATTERN})|(?:{_DOMAIN_PATTERN})", re.IGNORECASE
+    )
```
Comment (jon-betts) [inline h/views/badge.py:52]: Yeah this is real fun if you didn't spend 7 years doing Perl (we all have issues). The basic idea to make this maintainable at all is:

 * Declare all the parts that make this up separately "BLOCKED_DOMAINS" and "BLOCKED_SCHEMA"
 * Do all this horrible regex stuff
 * Test the hell out of it

File: h/views/badge.py [hunk 2]
```diff
@@ -22,25 +22,40 @@ def _has_uri_ever_been_annotated(db, uri):
...
 
     @classmethod
     def is_blocked(cls, url):
         """Check if a URL is blocked."""
 
-        return cls._DOMAIN_PATTERN.match(url)
+        return cls._PATTERN.match(url)
```
Comment (jon-betts) [inline h/views/badge.py:58]: And this is the pay off. It's super efficient as all the regex stuff will be done in C in one go.

File: tests/h/views/badge_test.py [hunk 3]
```diff
@@ -14,15 +14,22 @@ class TestBlocklist:
...
+        assert Blocklist.is_blocked(f"{scheme}://{suffix}")
+
     @pytest.mark.parametrize(
         "acceptable_url",
         (
             "http://example.com/this/is/fine",
             "http://example.com//facebook.com",
             "http://facebook.com.om.nom",
+            "file://c/my/magical_file.pdf",
+            "chrome-extension://blah",
```
Comment (jon-betts) [inline tests/h/views/badge_test.py:32]: Add some edge cases to make sure:

 * We aren't blocking everything that isn't `http*`
 * That we aren't blocking anything that just starts with `chrome`

File: h/views/badge.py (hunk 1 above)
Comment (lyzadanger) [reply h/views/badge.py:52]: My faaaaaace.

**Review by lyzadanger — APPROVED** [review]
Those regexes are _cruel_. But needs must?

**Comment by jon-betts** [comment]
An alternative to this is to use a Trie based method which can do efficient prefix matching. It's not quite as fast as this, but it's easier to grok.

I'll have a think about making this less horrible. This really is _very_ fast (1M/s), so we can probably afford quite a bit slower without it really having a big effect.

File: h/views/badge.py (hunk 1 above)
Comment (lyzadanger) [reply h/views/badge.py:66]: Thank you from the bottom of my glazed-over heart for those comments.

**Review by lyzadanger — APPROVED** [review]
While those comments might feel like "# set x to equal 4" to the author, they will be an invaluable asset to a future dev who would like to understand the intent of the regexes within the space of a few minutes.
